src/opera/parser/tosca/csar.py

```diff
@@ -1,31 +1,286 @@
+        if tmf_version not in CsarMeta.SUPPORTED_META_FILE_VERSIONS:
+            raise ParseError("{} {} is not supported. Supported versions: {}\".".format(
+                CsarMeta.KEY_TOSCA_META_FILE_VERSION,
+                tmf_version,
+                CsarMeta.SUPPORTED_META_FILE_VERSIONS
+            ), CsarMeta.METADATA_PATH)
+
+        result = CsarMeta(
+            name=parsed.get(CsarMeta.KEY_NAME),
+            content_type=parsed.get(CsarMeta.KEY_CONTENT_TYPE),
+            tosca_meta_file_version=tmf_version,
+            csar_version=csar_version,
+            created_by=parsed[CsarMeta.KEY_CREATED_BY],
+            entry_definitions=parsed.get(CsarMeta.KEY_ENTRY_DEFINITIONS),
+        )
+
+        return result
+
+    def render(self) -> str:
+        ordered_fields = [
+            (CsarMeta.KEY_NAME, self.name),
+            (CsarMeta.KEY_CONTENT_TYPE, self.content_type),
+            (CsarMeta.KEY_TOSCA_META_FILE_VERSION, self.tosca_meta_file_version),
+            (CsarMeta.KEY_CSAR_VERSION, self.csar_version),
+            (CsarMeta.KEY_CREATED_BY, self.created_by),
+            (CsarMeta.KEY_ENTRY_DEFINITIONS, self.entry_definitions)
+        ]
+
+        return "\n".join("{}: {}".format(k, v) for k, v in ordered_fields if v is not None)
+
+
+class ServiceTemplateMeta:
+    KEY_NAME = "template_name"
+    KEY_AUTHOR = "template_author"
+    KEY_VERSION = "template_version"
+
+    REQUIRED_KEYS = {KEY_NAME, KEY_VERSION}
+
+    def __init__(self, name: str, author: str, version: str):
+        self.name = name
+        self.author = author
+        self.version = version
+
+    @classmethod
+    def parse(cls, contents: str) -> "ServiceTemplateMeta":
+        parsed = yaml.safe_load(contents)
+
+        if "metadata" not in parsed:
+            raise ParseError("Missing required service template metadata.", "")
+
+        for req in ServiceTemplateMeta.REQUIRED_KEYS:
+            if req not in parsed["metadata"]:
+                raise ParseError("Missing required service template meta entry: metadata.{}".format(req), "")
+
+        return ServiceTemplateMeta(
+            name=parsed["metadata"][ServiceTemplateMeta.KEY_NAME],
+            author=parsed["metadata"].get(ServiceTemplateMeta.KEY_AUTHOR),
+            version=parsed["metadata"][ServiceTemplateMeta.KEY_VERSION],
+        )
+
+
+class CloudServiceArchive(ABC):
+    def __init__(self, csar_label):
+        self.csar_label = csar_label
+
+    @classmethod
+    def create(cls, path: PurePath) -> "CloudServiceArchive":
+        resolved_path = Path(path)
+
+        if not resolved_path.exists():
+            raise FileNotFoundError("File does not exist at {}".format(path))
+
+        if resolved_path.is_file():
+            return FileCloudServiceArchive(resolved_path)
+        elif resolved_path.is_dir():
+            return DirCloudServiceArchive(resolved_path)
+        else:
+            raise FileNotFoundError("CSARs are either regular files or directories.")
```

Comment:
Is FileNotFoundError appropriate here. Or is there some better error to describe the wrong CSAR type?

Reply:
It doesn't fit, true. I changed it to `OperaError` because it's related to opera, not the filesystem.